File: lib/src/tech_labs_library.dart
import 'models/notice_check_info.dart';
import 'models/notice_check_result.dart';
import 'models/server_environment.dart';
import 'models/service_info.dart';
import 'models/version_check_info.dart';
import 'models/version_check_result.dart';
import 'flutter_techlabs_library_platform_interface.dart';

class TechLabsLibrary {
  TechLabsLibrary._();

  static final TechLabsLibrary instance = TechLabsLibrary._();

  FlutterTechlabsLibraryPlatform get _platform =>
      FlutterTechlabsLibraryPlatform.instance;

  Future<void> configure({
    required String appId,
    required ServerEnvironment environment,
    int connectTimeoutMs = 10000,
    int readTimeoutMs = 10000,
  }) {
    return _platform.configure(
      appId: appId,
      environment: environment.value,
      connectTimeoutMs: connectTimeoutMs,
      readTimeoutMs: readTimeoutMs,
    );
  }

  Future<ServiceInfo> fetchServiceInfo() async {
    final map = await _platform.fetchServiceInfo();
    return ServiceInfo.fromMap(map);
  }

  Future<VersionCheckResult> checkVersion({
    VersionCheckInfo? info,
    String? currentVersion,
    int? currentBuildNumber,
  }) async {
    if (info == null) return const VersionCheckError();
    if (currentVersion == null && currentBuildNumber == null) {
      assert(false, 'checkVersion: currentVersion 또는 currentBuildNumber 중 하나는 필수');
      return const VersionCheckError();
    }
    final result = await _platform.checkVersion(
      info: info.toMap(),
      currentVersion: currentVersion,
      currentBuildNumber: currentBuildNumber,
    );
    return _parseVersionCheckResult(result);
  }

  Future<NoticeCheckResult> checkNotice({
    NoticeCheckInfo? info,
    required int lastSeenIndex,
  }) async {
    if (info == null) return const NoticeCheckError();
    final result = await _platform.checkNotice(
      info: info.toMap(),
      lastSeenIndex: lastSeenIndex,
    );
    return _parseNoticeCheckResult(result);
  }

  Future<void> clearCache() {
    return _platform.clearCache();
  }

  Future<bool> isUpdateRequired({
    String? serverVersion,
    String? currentVersion,
    int? serverBuildNumber,
    int? currentBuildNumber,
  }) {
    return _platform.isUpdateRequired(
      serverVersion: serverVersion,
      currentVersion: currentVersion,
      serverBuildNumber: serverBuildNumber,
      currentBuildNumber: currentBuildNumber,
    );
  }

  VersionCheckResult _parseVersionCheckResult(Map<Object?, Object?> map) {
    final resultStr = map['result'] as String?;
    final storePackage = (map['storePackage'] as String?) ?? '';
    switch (resultStr) {
      case 'forceUpdate':
        return ForceUpdate(storePackage);
      case 'optionalUpdate':
        return OptionalUpdate(storePackage);
      case 'noUpdateNeeded':
        return const NoUpdateNeeded();
      case 'error':
        return const VersionCheckError();
      default:
        assert(false, 'Unknown VersionCheckResult: $resultStr');
        return const VersionCheckError();
    }
  }

  NoticeCheckResult _parseNoticeCheckResult(Map<Object?, Object?> map) {
    final resultStr = map['result'] as String?;
    final latestIndex = map['latestIndex'] as int?;
    switch (resultStr) {
      case 'hasNewNotice':
        return HasNewNotice(latestIndex ?? 0);
      case 'noNewNotice':
        return const NoNewNotice();
      case 'error':
        return const NoticeCheckError();
      default:
        assert(false, 'Unknown NoticeCheckResult: $resultStr');
        return const NoticeCheckError();
    }
  }
}

File: android/src/main/kotlin/com/techlabs/flutter/flutter_techlabs_library/lib/TechLabsLibrary.kt
package com.techlabs.flutter.flutter_techlabs_library.lib

import com.techlabs.flutter.flutter_techlabs_library.lib.model.AppVersion
import com.techlabs.flutter.flutter_techlabs_library.lib.model.NoticeCheckInfo
import com.techlabs.flutter.flutter_techlabs_library.lib.model.NoticeCheckResult
import com.techlabs.flutter.flutter_techlabs_library.lib.model.ServiceInfo
import com.techlabs.flutter.flutter_techlabs_library.lib.model.VersionCheckInfo
import com.techlabs.flutter.flutter_techlabs_library.lib.model.VersionCheckResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

enum class ServerEnvironment {
    DEVELOPMENT,
    PRODUCTION
}

object TechLabsLibrary {

    @Volatile private var _appId: String? = null
    @Volatile private var _environment: ServerEnvironment? = null
    @Volatile private var _connectTimeoutMs: Int = 10_000
    @Volatile private var _readTimeoutMs: Int = 10_000
    @Volatile private var _serviceInfo: ServiceInfo? = null
    @Volatile private var _versionCheckInfo: VersionCheckInfo? = null

    private const val UPDATE_TYPE_FORCE = "0"
    private const val UPDATE_TYPE_OPTIONAL = "1"

    fun init(
        appId: String,
        environment: ServerEnvironment,
        connectTimeoutMs: Int = 10_000,
        readTimeoutMs: Int = 10_000
    ) {
        synchronized(this) {
            val changed = _appId != appId || _environment != environment
            _appId = appId
            _environment = environment
            _connectTimeoutMs = connectTimeoutMs
            _readTimeoutMs = readTimeoutMs
            if (changed) {
                _serviceInfo = null
                _versionCheckInfo = null
            }
        }
    }

    val appId: String
        get() = _appId ?: throw IllegalStateException("TechLabsLibrary is not initialized. Call init() first.")

    val s3Url: String?
        get() = synchronized(this) {
            val appId = _appId ?: throw IllegalStateException("TechLabsLibrary is not initialized.")
            val environment = _environment ?: throw IllegalStateException("TechLabsLibrary is not initialized.")
            getS3Url(appId, environment)
        }

    val serviceInfo: ServiceInfo? get() = _serviceInfo
    val versionCheckInfo: VersionCheckInfo? get() = _versionCheckInfo

    fun clearCache() {
        synchronized(this) {
            _serviceInfo = null
            _versionCheckInfo = null
        }
    }

    fun fetchServiceInfo(): ServiceInfo {
        val url = s3Url
            ?: throw IllegalStateException("No S3 URL registered for appId '${appId}'.")
        val json = fetchJson(url)
        val info = ServiceInfo.fromJsonString(json)
        synchronized(this) { _serviceInfo = info }
        return info
    }

    suspend fun fetchServiceInfoSuspend(): ServiceInfo {
        return withContext(Dispatchers.IO) { fetchServiceInfo() }
    }

    fun checkVersion(info: VersionCheckInfo?, currentBuildNumber: Int): VersionCheckResult {
        if (info == null) return VersionCheckResult.Error
        if (info.ver.isEmpty()) return VersionCheckResult.NoUpdateNeeded
        val serverBuildNumber = info.ver.toIntOrNull() ?: return VersionCheckResult.NoUpdateNeeded
        val isUpdateRequired = isUpdateRequired(serverBuildNumber, currentBuildNumber)
        return resolveResult(isUpdateRequired, info)
    }

    fun checkVersion(info: VersionCheckInfo?, currentVersion: String): VersionCheckResult {
        if (info == null) return VersionCheckResult.Error
        if (info.ver.isEmpty()) return VersionCheckResult.NoUpdateNeeded
        val isUpdateRequired = isUpdateRequired(info.ver, currentVersion)
        return resolveResult(isUpdateRequired, info)
    }

    fun checkVersion(infoMap: Map<String, Any?>?, currentBuildNumber: Int): VersionCheckResult {
        if (infoMap == null) return VersionCheckResult.Error
        return checkVersion(VersionCheckInfo.fromMap(infoMap), currentBuildNumber)
    }

    fun checkVersion(infoMap: Map<String, Any?>?, currentVersion: String): VersionCheckResult {
        if (infoMap == null) return VersionCheckResult.Error
        return checkVersion(VersionCheckInfo.fromMap(infoMap), currentVersion)
    }

    private fun resolveResult(isUpdateRequired: Boolean, info: VersionCheckInfo): VersionCheckResult {
        if (!isUpdateRequired) return VersionCheckResult.NoUpdateNeeded
        return when (info.type) {
            UPDATE_TYPE_FORCE -> VersionCheckResult.ForceUpdate(info.appName2)
            UPDATE_TYPE_OPTIONAL -> VersionCheckResult.OptionalUpdate(info.appName2)
            else -> VersionCheckResult.Error
        }
    }

    fun checkNotice(info: NoticeCheckInfo?, lastSeenIndex: Int): NoticeCheckResult {
        if (info == null) return NoticeCheckResult.Error
        val serverIndex = info.ntcLastIdx.toIntOrNull()
        if (info.ntcLastIdx.isEmpty() || serverIndex == null) return NoticeCheckResult.NoNewNotice
        return if (serverIndex > lastSeenIndex) NoticeCheckResult.HasNewNotice(serverIndex)
        else NoticeCheckResult.NoNewNotice
    }

    fun checkNotice(infoMap: Map<String, Any?>?, lastSeenIndex: Int): NoticeCheckResult {
        if (infoMap == null) return NoticeCheckResult.Error
        return checkNotice(NoticeCheckInfo.fromMap(infoMap), lastSeenIndex)
    }

    fun isUpdateRequired(serverVersion: String, currentVersion: String): Boolean {
        if (serverVersion.isBlank()) return false
        val server = AppVersion.fromString(serverVersion)
        val current = AppVersion.fromString(currentVersion)
        return server > current
    }

    fun isUpdateRequired(serverBuildNumber: Int, currentBuildNumber: Int): Boolean {
        return currentBuildNumber < serverBuildNumber
    }

    private fun fetchJson(url: String): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = _connectTimeoutMs
            readTimeout = _readTimeoutMs
        }
        try {
            val responseCode = connection.responseCode
            if (responseCode != HttpURLConnection.HTTP_OK) {
                throw IOException("Failed to fetch service info (HTTP $responseCode)")
            }
            return connection.inputStream.bufferedReader().use { it.readText() }
        } finally {
            connection.disconnect()
        }
    }

    private fun getS3Url(appId: String, environment: ServerEnvironment): String? {
        return when (appId) {
            "techlabs.global.yeppo" -> when (environment) {
                ServerEnvironment.DEVELOPMENT -> "https://status.yeppo.net/dev-yeppo-service.json"
                ServerEnvironment.PRODUCTION -> "https://status.yeppo.net/www-yeppo-service.json"
            }
            else -> null
        }
    }
}
